rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp

```diff
@@ -152,12 +152,19 @@ std::shared_ptr<rosbag2_storage::SerializedBagMessage> SequentialReader::read_ne
   throw std::runtime_error("Bag is not open. Call open() before reading.");
 }
 
+const rosbag2_storage::BagMetadata & SequentialReader::get_metadata() const
+{
+  return metadata_;
+}
+
 std::vector<rosbag2_storage::TopicMetadata> SequentialReader::get_all_topics_and_types()
 {
-  if (storage_) {
-    return storage_->get_all_topics_and_types();
+  topics_metadata_.clear();
+  topics_metadata_.reserve(metadata_.topics_with_message_count.size());
+  for (const auto & topic_information : metadata_.topics_with_message_count) {
+    topics_metadata_.push_back(topic_information.topic_metadata);
   }
-  throw std::runtime_error("Bag is not open. Call open() before reading.");
+  return topics_metadata_;
```

Comment:
I think we could do this operation once - when the metadata is read initially. Don't think we need to do that every time we ask for all topics given that this data shouldn't change.

Reply:
Yeah, you're right - changed to fill only once.